src/mame/namco/namcos10.cpp

```diff
@@ -1714,54 +2008,118 @@ void namcos10_memn_state::ns10_unks10md(machine_config &config)
 ///////////////////////////////////////////////////////////////////////////////////////////////
 // MEM(P3)
 
-void namcos10_memp3_state::nand_copy(uint8_t *nand_base, uint16_t *dst, uint32_t start_page, int len)
+void namcos10_memp3_state::firmware_write_w(uint16_t data)
 {
-	for (int page = start_page; page < start_page + len; page++)
-	{
-		int address = page * 0x210;
-
-		for (int i = 0; i < 0x200; i += 2) {
-			uint16_t data = nand_base[address + i + 1] | (nand_base[address + i] << 8);
-			*dst = m_unscrambler(data ^ 0xaaaa);
-			dst++;
-		}
+	if (data == 1) {
+		// This should happen in the device itself but it would require calling device_reset on the MCU
+		auto pc = ((m_mcu_ram[0x7fff02 / 2] << 16) | m_mcu_ram[0x7fff00 / 2]) & 0xffffff;
+		m_memp3_mcu->set_state_int(TLCS900_PC, pc);
+		m_memp3_mcu->resume(SUSPEND_REASON_HALT);
```

Comment:
You can’t call `->reset()` on the MCU?

Reply:
I thought the only way to do it was to call `device_reset()` which I remember getting told not to do before so I came up with something else to avoid it. This is cleaner and appears to work after a quick test, thanks.